### ISSUE TITLE
Automatic progress sync triggers hard hang on suspend

### ISSUE THREAD

**Issue by projectgus** [koreader/koreader#12614]
* KOReader version: master (92bcbe06f), nightly (`koreader-kobo-v2024.07-215-gd98fe14ba_2024-09-26.zip`), v2024.07 release
* Device: Kobo Aura

#### Issue

With "Automatically keep documents in sync" enabled and Wi-Fi connecting on demand for suspend, the Kobo often hard hangs on suspend and needs to be reset. 

#### Steps to reproduce

1. Configure a Wi-Fi Network.
2. Enable Progress Sync and "Automatically keep documents in sync" option.
3. Turn off Wi-Fi (or if "Disable Wi-Fi connection when inactive" is set then it seems enough to wait for connection to become inactive and disable itself.)
4. Suspend the Kobo by pressing the power button or closing the sleep cover.
5. Kobo observed to show screensaver and shows status info boxes for connecting to Wi-Fi ("Scanning for Network", "Wi-Fi Connected", etc.)
6. Kobo shows "Sleeping" info box on top of screensaver.
7. The bug is that from this point the Kobo often won't wake again, is unresponsive to the power button or cover opening. A hardware reset is required to bring it back to life (i.e. insert paperclip).

##### `crash.log` (if applicable)

[Verbose crash.log on current master](https://github.com/user-attachments/files/17306663/aura-crash-sync-suspend.log) - covers device booting, going to suspend via power button, Wi-Fi connecting, progress sync update, then going to suspend where it crashes. Log then starts again after hardware reset.

Happy to supply more logs if needed.

##### Other notes

* During this time the Kobo responds to pings over Wi-Fi for a window of ~20-30s, including for about 6-10 seconds after it shows "Sleeping" on the display, and then stops responding.
* A short time later my Wi-Fi AP logs "IEEE 802.11: disconnected due to excessive missing ACKs". If the Kobo doesn't hang then the AP log instead shows a clean `DHCPRELEASE` followed by `AP-STA-DISCONNECTED` event.
* Suspect the times it doesn't hang might all be when Wi-Fi is connected at the time it starts to suspend. Not sure about this, though.
* If the Kobo is left in the "hanging and unresponsive" state then the battery drains much faster than when actually suspended. It seems like it's either stuck awake and unresponsive, kernel has panicked, or similar.

Please let me know if I can provide any more information or test anything out.

**Comment by NiLuJe** [comment]
Given the device, likely to be a kernel issue we can't do much about, unfortunately.

(The Aura's kernel has historically been *extremely* buggy).

**Comment by NiLuJe** [comment]
Hmm, on that log, the sync was successful, which means it went to sleep with Wi-Fi enabled, which *does* implode the kernel on many older devices.

I'll double-check, because the order of events seems wonky here.

**Comment by projectgus** [comment]
That sounds right. It seems to me like:

* If Wi-Fi is on when the power button or close event happens, it syncs and then cleanly disconnects Wi-Fi and then suspends properly.
* If Wi-Fi is off when the power button or close events, it connects to WiFi and then syncs and then immediately goes to suspend and doesn't wake again. (Or possibly there's a race with going to suspend and sometimes it disconnects in time but sometimes it doesn't.)

**Comment by NiLuJe** [comment]
I haven't had time to check the PR in question, so feel free to have a go at it ;).

**Comment by globalhuman** [comment]
I have had a hard time developing a feature for the statistics plugin to allow auto sync similar to how reading progress is synced i.e. `onSuspend`, `onCloseDocument`, `onNetworkDisconnected`, `onNetworkDisconnecting`.

I was facing this same issue on a kobo libra. Being new to this repo I suspected it was likely a change that I wrote, but will test out these changes and will report back if they help.
